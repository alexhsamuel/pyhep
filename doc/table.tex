\section{\module{hep.table} --- Tables}

\declaremodule{extension}{hep.table}
\modulesynopsis{Tables.}

A \pyhep \emph{table} is a flat database table with columns of fixed
type and storage size.  Tables are append-only: new rows may be added to
the end of the table, but rows may not be inserted elsewhere, nor can
rows be modified or deleted.  

Each table has a fixed \emph{schema}, which determines the data items
stored in the table and their types.  The schema includes an unordered
set of \emph{columns}, each identified by a name.  Each column has a
fixed data type.  A schema may also include \emph{expressions}, which
are derived quantities which are computed for each row from values of
columns and other expressions for that row.  

The table's data is an ordered sequence of \emph{rows}, each of which
contains exactly one value for each column in the table's schema.

Multiple table implementations are provided.  Implementations provide
different interfaces for creating and opening tables.  However, objects
representing tables and rows satisfy the same protocol across all
implementations.


\subsection{Schema objects}

An instance of \class{hep.table.Schema} represents a table schema.  

\begin{memberdesc}{columns}
 \readonly A sequence of columns in the schema.  Each element is a
 \class{Column} instance.  The order of the columns is unspecified.
\end{memberdesc}

\begin{memberdesc}{expressions}
 \readonly A sequence of expressions in the schema.  Each element is
 expression object (see \module{hep.expr}).  The order of the
 expressions is unspecified.
\end{memberdesc}

\begin{methoddesc}{addColumn}{name, type\optional{, **attributes}}
 Add a column to the sequence.  \var{name} is a character string, and
 \var{type} is a column type (see below).  If additional keyword
 arguments are provided, they are added to the column as attributes.
 New columns should not be added to a schema after the schema has been
 used to create a table.
\end{methoddesc}

\begin{methoddesc}{addExpression}{name, expression\optional{, **attributes}}
 Add an expression to the sequence.  \var{name} is a character string,
 and \var{expression} is an expression as a string or expression object.
 If additional keyword arguments are provided, they are added to the
 expression object as attributes.
\end{methoddesc}

An instance of \class{hep.table.Column} describes a column in a schema.
Do not instantiate this class directly; instead, use the
\method{addColumn} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The column's name.
\end{memberdesc}

\begin{memberdesc}{type}
 \readonly The column's data type.
\end{memberdesc}

The column's data type is specified by one of these names:
\begin{tableiii}{ccc}{code}{name}{C type}{Python Type}
  \lineiii{"int8"}{\code{signed char}}{\code{int}}
  \lineiii{"int16"}{\code{signed short}}{\code{int}}
  \lineiii{"int32"}{\code{signed int}}{\code{int}}
  \lineiii{"float32"}{\code{float}}{\code{float}}
  \lineiii{"float64"}{\code{double}}{\code{float}}
\end{tableiii}


An instance of \class{hep.table.Expression} describes a column in a
schema.  Do not instantiate this class directly; instead, use the
\method{addExpression} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The expressions's name.
\end{memberdesc}

\begin{memberdesc}{expression}
 \readonly An \class{Expression} instance; see \module{hep.expr}.
\end{memberdesc}

\begin{memberdesc}{formula}
 \readonly A string representation of the expression.
\end{memberdesc}



\subsection{Table objects}

A table object represents an in-memory table or a connection to an
externally-stored table.  In the former case, the table is erased when
the table object is deleted.  In the latter case, the connection is
closed but the table persists when the table object is deleted.

A table object satisfies the Python read-only sequence protocol.  The
sequence elements are the rows of the sequence.  Thus, the
\function{len} function returns the number of rows in the table, and
the \method{__getitem__} method extracts a row by its position in the
sequence. 

In addition, these methods are provided:

\begin{methoddesc}{append}{row\optional{, **kw_args}}
 (Read-write tables only.)  Appends a row to the end of the table.  The
 \var{row} must is a dictionary or other mapping object which maps
 column names to values.  Additional column values may be specified as
 keyword arguments.  If a column is specified both in \var{row} and in
 \var{**kw_args}, the value from latter is used.  A value must be
 specified for each column, and no other names may be specified.
 Returns the row index of the new row.
\end{methoddesc}

\begin{methoddesc}{__iter__}{}
 Returns an iterator over all rows in the table.
\end{methoddesc}

\begin{memberdesc}{rows}
 An interator over all rows in the table.
\end{memberdesc}

\begin{methoddesc}{select}{expr}
 Returns an iterator over rows in the table for which expression
 \var{expr} is true.  \var{expr} may be a string expression formula or
 an expression object.
\end{methoddesc}


\subsection{Row objects}

A row object may represent a row in a table, or a row that is compatible
with a table's schema (but not one of the rows in that table).

The row object satisfies the Python map protocol.  Each key is the name
of a column in the table's schema.  The corresponding map value is the
row's value for that column.  Row objects should be treated as
read-only; it is not possible to modify a table row by changing values
in a row object.

In addition, a row object provides these members and methods:

\begin{methoddesc}{asDict}{}
 Return a new Python dictionary of the column names and corresponding
 values in the row.
\end{methoddesc}

\begin{memberdesc}{index}
 \readonly The index of the row in the table.  The value of this
 attribute is undefined if the row was obtained from the table's
 \method{newRow} method and has not yet been passed to \method{append}.
\end{memberdesc}

\begin{memberdesc}{table}
 \readonly The table to which the row belongs.
\end{memberdesc}


\subsection{Standard implementation}

\pyhep's standard table implementation, in \module{hep.table}, uses a
simple, binary, on-disk representation of the table's data.  Table rows
are automatically loaded from disk as needed.

The \module{hep.table} module provides functions \function{create} and
\function{open} to create and open new tables, respectively.

\begin{funcdesc}{create}{filename, schema}
 Create a new table.  The table is stored in a file named by
 \var{filename}, which is created or overwritten.  The \var{schema}
 argument is a map from column names to \class{Column} instances,
 specifying the columns in the table.

 The return value is a table object, which is opened to the new, empty
 in write mode.
\end{funcdesc}

\begin{funcdesc}{open}{filename\optional{, mode="r"}}
 Open an existing table stored in the file named by \var{filename}.  The
 \var{mode} argument specifies the mode in which to open the table:
 \constant{"r"} to open the table in read-only mode, or \constant{"w"}
 to open the table in read-write mode.

 The return value is a table object.
\end{funcdesc}

