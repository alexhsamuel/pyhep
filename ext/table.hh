//----------------------------------------------------------------------
//
// table.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Flat-file data table implementation and file format.  */

#ifndef __TABLE_HH__
#define __TABLE_HH__

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <fcntl.h>
#include <string>
#include <sys/types.h>
#include <unistd.h>
#include <vector>

#include "value.hh"

//----------------------------------------------------------------------
// forward declarations
//----------------------------------------------------------------------

namespace table {

class Row;

//----------------------------------------------------------------------
// exception classes
//----------------------------------------------------------------------

class NoColumn 
{
public:

  NoColumn(const std::string& column_name);
  const std::string column_name_;

};


inline
NoColumn::NoColumn(const std::string& column_name)
  : column_name_(column_name)
{
}


class NotWritable
{
};


class NoRow
{
public:

  NoRow(int row_number) : row_number_(row_number) {}
  const int row_number_;

};


class FileError
{
public:

  FileError(const std::string& message);
  const std::string message_;

};


inline
FileError::FileError(const std::string& message)
  : message_(message)
{
}


class WrongColumnType
{
};


//----------------------------------------------------------------------
// types
//----------------------------------------------------------------------

enum ColumnType
{
  TYPE_NONE = -1,
  TYPE_BOOL = 0,
  TYPE_INT_8,
  TYPE_INT_16,
  TYPE_INT_32,
  TYPE_FLOAT_32,
  TYPE_FLOAT_64,
  TYPE_COMPLEX_64,
  TYPE_COMPLEX_128,
  TYPE_LAST
};


//----------------------------------------------------------------------

#if 0
template<typename TYPE>
struct ColumnTypeTemplate
{
  static ColumnType type;
};


template<>
ColumnType
ColumnTypeTemplate<bool>::type = TYPE_BOOL;


template<>
ColumnType
ColumnTypeTemplate<char>::type = TYPE_INT_8;


template<>
ColumnType
ColumnTypeTemplate<short>::type = TYPE_INT_16;


template<>
ColumnType
ColumnTypeTemplate<long>::type = TYPE_INT_32;


template<>
ColumnType
ColumnTypeTemplate<float>::type = TYPE_FLOAT_32;


template<>
ColumnType
ColumnTypeTemplate<double>::type = TYPE_FLOAT_64;
#else

template<typename TYPE>
ColumnType
getColumnType();

template<>
inline ColumnType
getColumnType<bool>()
{
  return TYPE_BOOL;
}


template<>
inline ColumnType
getColumnType<char>()
{
  return TYPE_INT_8;
}


template<>
inline ColumnType
getColumnType<short>()
{
  return TYPE_INT_16;
}


template<>
inline ColumnType
getColumnType<long>()
{
  return TYPE_INT_32;
}


template<>
inline ColumnType
getColumnType<float>()
{
  return TYPE_FLOAT_32;
}


template<>
inline ColumnType
getColumnType<double>()
{
  return TYPE_FLOAT_64;
}
#endif


//----------------------------------------------------------------------

class Column
{
public:

  Column(const std::string& name, const ColumnType type);
  Column(const Column& column);

  const std::string& getName() const;
  ColumnType getType() const;

private:

  std::string name_;
  ColumnType type_;

};


//----------------------------------------------------------------------

class Schema
{
public:

  Schema();
  virtual ~Schema();

  int getNumColumns() const;
  int whichColumn(const std::string& name) const throw (NoColumn);
  const Column& getColumn(int column_index) const;
  size_t getColumnOffset(int column_index) const;

  size_t getSize() const;

protected:

  struct ColumnRecord 
  {
    Column column_;
    size_t offset_;
  };

  std::vector<ColumnRecord> columns_;
  size_t size_;

};


class CustomSchema
  : public Schema
{
public:

  CustomSchema(size_t size=0);
  virtual ~CustomSchema();

  void addColumn(const Column& column, size_t offset);

};


class AutoSchema
  : public Schema
{
public:

  AutoSchema();
  virtual ~AutoSchema();

  void addColumn(const Column& column);

};


// FIXME: This is disabled; is it worth fixing?
#if 0
template<class STRUCT>
class StructSchema
  : public Schema
{
public:

  StructSchema();
  virtual ~StructSchema();

  template<typename FIELD_TYPE>
  void addColumn(const std::string& name, FIELD_TYPE STRUCT::* offset);

  void fillRow(Row* row, const STRUCT* structure) const;

private:

  void addColumn(const std::string& name, int offset, ColumnType type);

  std::vector<int> offsets_;

};
#endif


class Row
{
public:

  Row(const Schema* schema);
  ~Row();

  const Schema* getSchema() const
    { return schema_; }

  char* getBuffer() const
    { return data_; }

  Value getValue(int column_index) const throw (WrongColumnType);
  void setValue(int column_index, const Value& value);

private:

  const Schema* const schema_;
  char* data_;

};


class Table
{
protected:

  Table(const Schema* schema, bool writable)
    : schema_(schema), writable_(writable) {}

public:

  virtual ~Table();

  const Schema* getSchema() const
    { return schema_; }
  const bool isWritable() const
    { return writable_; }
  
  virtual int64_t getNumRows() const = 0;
  virtual void read(int64_t row_number, Row* row) = 0;
  virtual int append(const Row* row) = 0;
  virtual std::string getMetadata() const = 0;
  virtual void setMetadata(const std::string& data) = 0;

protected:

  const Schema* schema_;
  bool writable_;

};


class FileTable
  : public Table
{
public:

  ~FileTable();

  virtual int64_t getNumRows() const;
  virtual void read(int64_t row_number, Row* row);
  virtual int append(const Row* row);
  virtual std::string getMetadata() const;
  virtual void setMetadata(const std::string& data);

  static FileTable* create(const Schema* schema,
			   const std::string& path, 
			   mode_t mode=0666);
  static FileTable* open(const std::string& path, 
			 const std::string& mode);
  std::string getPath() const
    { return path_; }

protected:

  FileTable(const std::string& path, int flags=O_RDONLY);

private:

  const std::string path_;
  int fd_;

  off64_t first_row_offset_;
  off64_t row_size_;

  int64_t num_rows_;

};


//----------------------------------------------------------------------
// function declarations
//----------------------------------------------------------------------

extern size_t
getTypeSize(ColumnType type);

extern const char*
getTypeName(ColumnType type);

extern ColumnType
getTypeByName(const char* name);

//----------------------------------------------------------------------
// inline methods
//----------------------------------------------------------------------

inline
Column::Column(const std::string& name,
	       const ColumnType type)
  : name_(name),
    type_(type)
{
}


inline
Column::Column(const Column& column)
  : name_(column.name_),
    type_(column.type_)
{
}


inline const std::string&
Column::getName()
  const
{
  return name_;
}


inline ColumnType
Column::getType()
  const
{
  return type_;
}


//----------------------------------------------------------------------

inline
Schema::Schema()
  : size_(0)
{
}


inline int
Schema::getNumColumns()
  const
{
  return columns_.size();
}


inline const Column&
Schema::getColumn(int column_index)
  const
{
  return columns_[column_index].column_;
}


inline size_t
Schema::getColumnOffset(int column_index)
  const
{
  return columns_[column_index].offset_;
}


inline size_t
Schema::getSize()
  const
{
  return size_;
}


//----------------------------------------------------------------------

inline
CustomSchema::CustomSchema(size_t size)
{
  size_ = size;
}


inline
CustomSchema::~CustomSchema()
{
}


inline void
CustomSchema::addColumn(const Column& column,
			size_t offset)
{
  ColumnRecord record = { column, offset };
  columns_.push_back(record);
}


//----------------------------------------------------------------------

#if 0
template<class STRUCT>
inline
StructSchema<STRUCT>::StructSchema()
{
}


template<class STRUCT>
inline
StructSchema<STRUCT>::~StructSchema()
{
}


template<class STRUCT>
template<typename FIELD_TYPE>
inline void
StructSchema<STRUCT>::addColumn(const std::string& name,
				FIELD_TYPE STRUCT::* offset)
{
  addColumn(name, (int) &(((STRUCT*) NULL)->*offset),
	    getColumnType<FIELD_TYPE>());
}


template<class STRUCT>
inline void
StructSchema<STRUCT>::fillRow(Row* row,
			      const STRUCT* structure)
  const
{
  int num_columns = columns_.size();
  for (int c = 0; c < num_columns; ++c) {
    const char* field_address = ((const char*) structure) + offsets_[c];
    row->setValue(c, (void*) field_address);
  }
}


template<class STRUCT>
inline void
StructSchema<STRUCT>::addColumn(const std::string& name,
				int offset,
				ColumnType type)
{
  Column column(name, type);
  ColumnRecord record = { column, size_ };
  columns_.push_back(record);
  size_ += getTypeSize(column.getType());
  offsets_.push_back(offset);
}
#endif


//----------------------------------------------------------------------

inline int64_t
FileTable::getNumRows()
  const
{
  return num_rows_;
}


//----------------------------------------------------------------------

}  // namespace table


#endif  // #ifndef __TABLE_HH__
