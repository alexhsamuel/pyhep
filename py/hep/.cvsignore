*.pyc
*.pyo
config.py
*-install
pdt.pickle
