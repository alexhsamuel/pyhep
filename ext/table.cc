//----------------------------------------------------------------------
//
// table.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cerrno>
#include <complex>
#include <fcntl.h>
#include <libgen.h>
#include <stdint.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "table.hh"
#include "value.hh"

namespace table {

namespace {

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

const size_t
type_sizes[] = { 
  1,    // TYPE_BOOL
  1,    // TYPE_INT_8
  2,    // TYPE_INT_16
  4,    // TYPE_INT_32
  4,    // TYPE_FLOAT_32
  8,    // TYPE_FLOAT_64
  8,    // TYPE_COMPLEX_64
  16,   // TYPE_COMPLEX_128
 };


const char* const
type_names[] = {
  "bool",       // TYPE_BOOL
  "int8",       // TYPE_INT_8
  "int16",      // TYPE_INT_16
  "int32",      // TYPE_INT_32
  "float32",    // TYPE_FLOAT_32
  "float64",    // TYPE_FLOAT_64
  "complex64",  // TYPE_COMPLEX_64
  "complex128", // TYPE_COMPLEX_128
};


const int
file_format_magic_number = 0x11a66826;

const int
file_format_version_number = 6;


//----------------------------------------------------------------------
// private types
//----------------------------------------------------------------------

// We have stdint.h to provide int32_t etc. but nothing for similar
// floating-point types.  Define the ones we need here.

// FIXME: These need to be configured specifically for each
// architecture. 

typedef float
float32_t;

typedef double
float64_t;

typedef std::complex<float>
complex64_t;

typedef std::complex<double>
complex128_t;

//----------------------------------------------------------------------

struct FileFormatHeader
{
  /* A magic number, identifying this file as a table file.  */
  int magic_number_;

  /* The file format version number.  */
  int version_number_;

  /* The number of rows in the table.  */
  int num_rows_;

  /* The offset to the data for the first row.  */
  off64_t first_row_offset_;

  /* Ignored, but retained for file format compatibility.  */
  off64_t metadata_offset_;

  /* Ignored, but retained for file format compatibility.  */
  off64_t metadata_len_;
};


//----------------------------------------------------------------------

class Buffer 
{
public:

  Buffer(size_t size) : pointer_(new char[size]) {}
  ~Buffer() { delete[] pointer_; }

  char* const pointer_;

};


//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline size_t
xwrite(int fd,
       const void* buffer,
       size_t count)
  throw (FileError)
{
  size_t result = ::write(fd, buffer, count);
  if (result != count)
    throw FileError(strerror(errno));
  return count;
}


inline void
xread(int fd, 
      void* buffer,
      size_t count)
  throw (FileError)
{
  size_t result = ::read(fd, buffer, count);
  if (result != count)
    throw FileError(strerror(errno));
}


inline void
xseek(int fd,
      off64_t offset)
  throw (FileError)
{
  off64_t result = ::lseek64(fd, offset, SEEK_SET);
  if (result != offset)
    throw FileError(strerror(errno));
}


inline std::string
getMetadataPath(const std::string& path)
{
  return path + ".metadata";
}


size_t
writeSchema(int fd, 
	    const Schema* schema)
{
  assert(fd >= 0);
  assert(schema != NULL);
  size_t written = 0;

  size_t size = schema->getSize();
  written += xwrite(fd, &size, sizeof(size_t));

  int num_columns = schema->getNumColumns();
  written += xwrite(fd, &num_columns, sizeof(int));

  for (int c = 0; c < num_columns; ++c) {
    const Column& column = schema->getColumn(c);

    ColumnType type = column.getType();
    written += xwrite(fd, &type, sizeof(ColumnType));

    std::string name = column.getName();
    int name_length = name.length();
    const char* name_string = name.data();
    written += xwrite(fd, &name_length, sizeof(int));
    written += xwrite(fd, name_string, name_length);

    size_t offset = schema->getColumnOffset(c);
    written += xwrite(fd, &offset, sizeof(offset));
  }

  return written;
}


Schema*
readSchema(int fd)
{
  assert(fd >= 0);
  
  size_t size;
  xread(fd, &size, sizeof(size_t));

  CustomSchema* schema = new CustomSchema(size);

  int num_columns;
  xread(fd, &num_columns, sizeof(int));

  for (int c = 0; c < num_columns; ++c) {
    ColumnType type;
    xread(fd, &type, sizeof(ColumnType));

    int name_length;
    xread(fd, &name_length, sizeof(int));
    char name[name_length + 1];
    xread(fd, name, name_length);
    name[name_length] = '\0';

    size_t offset;
    xread(fd, &offset, sizeof(offset));

    schema->addColumn(Column(name, type), offset);
  }

  return schema;
}


}  // anonymous namespace


//----------------------------------------------------------------------
// class Row
//----------------------------------------------------------------------

Row::Row(const Schema* schema)
  : schema_(schema)
{
  assert(schema_ != 0);
  data_ = new char[schema_->getSize()];
}


Row::~Row()
{
  delete [] data_;
}


#define ROW_GET(TYPE, OFFSET) \
  (*((TYPE*) (data_ + (OFFSET))))

Value
Row::getValue(const int column_index)
  const
  throw (WrongColumnType)
{
  ColumnType type = schema_->getColumn(column_index).getType();
  size_t offset = schema_->getColumnOffset(column_index);
  switch (type) {
  case TYPE_NONE:
    abort();

  case TYPE_BOOL:
    return Value::make((bool) ROW_GET(int8_t, offset));
		       
  case TYPE_INT_8:
    return Value::make((long) ROW_GET(int8_t, offset));

  case TYPE_INT_16:
    return Value::make((long) ROW_GET(int16_t, offset));
    
  case TYPE_INT_32:
    return Value::make((long) ROW_GET(int32_t, offset));
    
  case TYPE_FLOAT_32:
    return Value::make((double) ROW_GET(float32_t, offset));

  case TYPE_FLOAT_64:
    return Value::make((double) ROW_GET(float64_t, offset));

  case TYPE_COMPLEX_64: 
    return Value::make((std::complex<double>) ROW_GET(complex64_t, offset));

  case TYPE_COMPLEX_128: 
    return Value::make((std::complex<double>) ROW_GET(complex128_t, offset));

  default:
    abort();
  }
}


#define ROW_SET(TYPE, OFFSET, VALUE) \
  *((TYPE*) (data_ + (OFFSET))) = ((TYPE) (VALUE))

void
Row::setValue(const int column_index,
	      const Value& value)
{
  assert(column_index >= 0 && column_index < schema_->getNumColumns());

  ColumnType type = schema_->getColumn(column_index).getType();
  size_t offset = schema_->getColumnOffset(column_index);
  switch (type) {
  case TYPE_NONE:
    abort();

  case TYPE_BOOL:
    ROW_SET(int8_t, offset, value.as_bool());
    break;
		       
  case TYPE_INT_8:
    ROW_SET(int8_t, offset, value.as_long());
    break;

  case TYPE_INT_16:
    ROW_SET(int16_t, offset, value.as_long());
    break;
    
  case TYPE_INT_32:
    ROW_SET(int32_t, offset, value.as_long());
    break;
    
  case TYPE_FLOAT_32:
    ROW_SET(float32_t, offset, value.as_double());
    break;

  case TYPE_FLOAT_64:
    ROW_SET(float64_t, offset, value.as_double());
    break;

  case TYPE_COMPLEX_64:
    ROW_SET(complex64_t, offset, value.as_complex());
    break;

  case TYPE_COMPLEX_128:
    ROW_SET(complex128_t, offset, value.as_complex());
    break;

  default:
    abort();
  }
}  


//----------------------------------------------------------------------
// class Schema
//----------------------------------------------------------------------

Schema::~Schema()
{
}


int
Schema::whichColumn(const std::string& name)
  const
  throw (NoColumn)
{
  int num_columns = columns_.size();
  for (int i = 0; i < num_columns; ++i) 
    if (columns_[i].column_.getName() == name)
      return i;

  throw NoColumn(name);
}


//----------------------------------------------------------------------
// class AutoSchema
//----------------------------------------------------------------------

AutoSchema::AutoSchema()
{
}


AutoSchema::~AutoSchema()
{
}


void
AutoSchema::addColumn(const Column& column)
{
  ColumnRecord record = { column, size_ };
  columns_.push_back(record);
  size_ += getTypeSize(column.getType());
}


//----------------------------------------------------------------------
// class Table
//----------------------------------------------------------------------

Table::~Table()
{
}


//----------------------------------------------------------------------
// class FileTable
//----------------------------------------------------------------------

FileTable*
FileTable::create(const Schema* schema,
		  const std::string& path,
		  mode_t mode)
{
  // Create the file.
  int fd = ::open64(path.c_str(), 
		    O_RDWR | O_CREAT | O_TRUNC | O_LARGEFILE, mode);
  if (fd < 0)
    throw FileError(strerror(errno));

  // Skip forward to leave enough room for the header, then write the
  // schema.
  xseek(fd, sizeof(FileFormatHeader));
  off64_t schema_size = writeSchema(fd, schema);

  // Construct the header info, now that we know how large the schema
  // is.
  FileFormatHeader header;
  header.magic_number_ = file_format_magic_number;
  header.version_number_ = file_format_version_number;
  header.num_rows_ = 0;
  header.first_row_offset_ = sizeof(FileFormatHeader) + schema_size;

  // Write the header at the beginning of the file.
  xseek(fd, 0);
  xwrite(fd, &header, sizeof(header));

  // Close the file.
  close(fd);

  // If there is a metadata file for this path, remove it.
  std::string metadata_path = getMetadataPath(path);
  if (::access(metadata_path.c_str(), X_OK))
    ::unlink(metadata_path.c_str());

  // Now open the newly-created table in the usual way.
  return new FileTable(path, O_RDWR);
}


FileTable*
FileTable::open(const std::string& path,
		const std::string& mode)
{
  int flags;
  if (mode == "r") 
    flags = O_RDONLY;
  else if (mode == "w") 
    flags = O_RDWR;
  else
    throw FileError("unknown mode");

  return new FileTable(path, flags);
}


FileTable::~FileTable()
{
  if (isWritable()) {
    // Build the header.
    FileFormatHeader header;
    header.magic_number_ = file_format_magic_number;
    header.version_number_ = file_format_version_number;
    header.num_rows_ = num_rows_;
    header.first_row_offset_ = first_row_offset_;

    // Write the header.
    xseek(fd_, 0);
    xwrite(fd_, &header, sizeof(header));

    int result = fsync(fd_);
    if (result != 0)
      throw FileError(strerror(errno));
  }

  int result;
  result = close(fd_);
  assert(result == 0);
}


void
FileTable::read(int64_t row_number, 
		Row* row)
{
  assert(row->getSchema() == this->getSchema());
  assert(row_number >= 0 && row_number < num_rows_);

  off64_t offset = first_row_offset_ + row_number * row_size_;
  xseek(fd_, offset);
  xread(fd_, row->getBuffer(), row_size_);
}


int
FileTable::append(const Row* row)
{
  assert(row->getSchema() == this->getSchema());

  if (! isWritable())
    throw NotWritable();

  int64_t row_number = num_rows_;
  off64_t offset = first_row_offset_ + row_number * row_size_;
  xseek(fd_, offset);
  xwrite(fd_, row->getBuffer(), row_size_);

  ++num_rows_;
  return row_number;
}


std::string
FileTable::getMetadata()
  const
{
  std::string path = getMetadataPath(path_);

  // Can we access the metadata file?
  if (::access(path.c_str(), R_OK) != 0)
    // Can't access the metadata file.  Either it doesn't exist or it's
    // inaccessible.  In either case, just return empty metadata.
    return "";

  // Get file information.
  struct stat file_info;
  if (::stat(path.c_str(), &file_info) != 0)
    return "";
  if (file_info.st_size == 0)
    // No metadata.
    return "";

  // Open the metadata file.
  int fd = ::open(path.c_str(), O_RDONLY);
  if (fd == -1) 
    return "";

  // Construct a buffer to hold the metadata.
  std::auto_ptr<Buffer> buffer(new Buffer(file_info.st_size));

  // Read the data.
  xread(fd, buffer->pointer_, file_info.st_size);
  // Clean up.
  ::close(fd);

  return std::string(buffer->pointer_, file_info.st_size);
}


void
FileTable::setMetadata(const std::string& data)
{
  std::string path = getMetadataPath(path_);

  if (data == "") {
    // We don't want a metadata file.  Remove one, if it exists.
    if (::access(path.c_str(), X_OK))
      ::unlink(path.c_str());
    // Don't write.
    return;
  }

  // Create the metadata file.
  int fd = ::open(path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0666);
  if (fd == -1) 
    // Could not open the metadata file; fail silently.
    return;

  // Write the metadata.
  xwrite(fd, data.c_str(), data.length());
  // Clean up.
  close(fd);
}


FileTable::FileTable(const std::string& path,
		     int flags)
  : Table(NULL, flags == O_RDWR),
    path_(path)
{
  if (flags != O_RDWR && flags != O_RDONLY)
    // Invalid flags.
    assert(false);

  // Open the file.
  fd_ = ::open64(path.c_str(), flags | O_LARGEFILE);
  if (fd_ < 0) {
    throw FileError(strerror(errno));
  }

  // Read the header.
  FileFormatHeader header;
  xread(fd_, &header, sizeof(header));

  // Check the magic numbers.
  if (header.magic_number_ != file_format_magic_number)
    throw FileError("wrong file format");
  if (header.version_number_ != file_format_version_number)
    throw FileError("wrong file format version");

  // Extract other info from the header.
  num_rows_ = header.num_rows_;
  assert(num_rows_ >= 0);
  first_row_offset_ = header.first_row_offset_;

  // Read the schema.
  schema_ = readSchema(fd_);
  assert(schema_ != NULL);
  row_size_ = schema_->getSize();
}


//----------------------------------------------------------------------
// function definitions
//----------------------------------------------------------------------

size_t
getTypeSize(ColumnType type)
{
  return type_sizes[(int) type];
}


const char*
getTypeName(ColumnType type)
{
  int i = (int) type;
  if (i >= 0 && i < (int) TYPE_LAST)
    return type_names[i];
  else
    return NULL;
}


ColumnType
getTypeByName(const char* name)
{
  // Try to match the name.
  for (int i = 0; i < (int) TYPE_LAST; ++i)
    if (strcmp(name, type_names[i]) == 0)
      return (ColumnType) i;
  // No match.
  return TYPE_NONE;
}


//----------------------------------------------------------------------

}  // namespace Tables
