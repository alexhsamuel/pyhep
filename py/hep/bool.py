#-----------------------------------------------------------------------
#
# module hep.bool
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Implementation of bool for older Python versions.

This module provides 'bool', 'True', and 'False' similar to those found
in recent versions of Python.  These symbols are defined only when
Python doesn't include its own built-in."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.ext import BoolArray

#-----------------------------------------------------------------------

if not hasattr(__builtins__, "bool"):

    False = 0

    True = 1


    class bool(object):

        def __new__(class_, value):
            if value:
                return True
            else:
                return False



