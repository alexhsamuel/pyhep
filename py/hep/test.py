#-----------------------------------------------------------------------
#
# module hep.test
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Regression testing support."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.bool import *
import sys

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TestFailure(Exception):
    pass



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

def assert_(expression):
    if not expression:
        raise TestFailure, "assertion failed"


def compare(value, expected, precision=None):
    if type(value) in (list, tuple) and type(expected) in (list, tuple):
        if len(value) != len(expected):
            result = False
        else:
            try:
                result = reduce(
                    lambda r1, r2: r1 and r2,
                    map(lambda (v, e): compare(v, e, precision),
                        zip(value, expected)),
                    True)
            except TestFailure:
                result = False

    elif precision is not None:
        result = (abs(value - expected) <= precision)

    else:
        result = (value == expected)

    if result:
        return True
    else:
        raise TestFailure, \
              "expected %s; got %s" % (repr(expected), repr(value))


def compareSequence(value, expected, ordered=True):
    value = list(value)
    expected = list(expected)
    if not ordered:
        value.sort()
        expected.sort()
    return compare(value, expected)


