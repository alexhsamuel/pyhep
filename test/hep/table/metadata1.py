#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.table
from   hep.test import compare

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

schema = hep.table.Schema()
schema.addColumn("count", "int16", title="number of grapefruits")
schema.addColumn("length", "float32", units="meters")
schema.name = "sample schema"

table = hep.table.create("metadata1.table", schema)
table.append(count=10, length=0.65)

del table, schema

#-----------------------------------------------------------------------

table = hep.table.open("metadata1.table")
schema = table.schema
print schema

compare(len(schema), 2)
compare(schema["count"].title, "number of grapefruits")
compare(schema["length"].units, "meters")
compare(schema.name, "sample schema")
compare(len(table), 1)

del table, schema

#-----------------------------------------------------------------------

# Overwrite the table with another by the same name but a different
# schema, to make sure the old metadata file gets clobbered and doesn't
# conflict with the new table.

schema = hep.table.Schema()
schema.addColumn("x", "int32", title="stuff")
schema.name = "sample schema"

table = hep.table.create("metadata1.table", schema)
table.append(x=42)
table.append(x=39)

del table, schema

#-----------------------------------------------------------------------

table = hep.table.open("metadata1.table")
schema = table.schema
print schema

compare(len(schema), 1)
compare(schema["x"].title, "stuff")
compare(len(table), 2)

del table, schema

