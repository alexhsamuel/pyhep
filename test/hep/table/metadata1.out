Schema(count='int16', length='float32')
Schema(x='int32')
